R/brace_linter.R

```diff
@@ -8,9 +8,14 @@
 #'  - Closing curly braces in `if` conditions are on the same line as the corresponding `else`.
 #'  - Either both or neither branch in `if`/`else` use curly braces, i.e., either both branches use `{...}` or neither
 #'    does.
-#'  - Functions spanning multiple lines use curly braces.
+#'  - Function bodies are wrapped in curly braces.
 #'
-#' @param allow_single_line if `TRUE`, allow an open and closed curly pair on the same line.
+#' @param allow_single_line If `TRUE`, allow an open and closed curly pair on the same line.
+#' @param function_braces Character specifying whether to require function bodies to be wrapped in curly braces.
+#'   `"always"` requires braces for all function definitions, including inline functions.
+#'   `"not_inline"` requires braces when a function body does not start on the same line as its header.
+#'   `"multi_line"` requires braces when a function definition spans multiple lines.
+#'   `"never"` never requires braces in function bodies.
```

Comment:
we don't have to include it in this PR, but there is one more style we might cater to in the future, namely, use braces _if and only if_ they're syntactically required.

Consider

```r
A <- function() { NULL }
B <- function() { NULL; NULL }
C <- function() NULL
D <- function() {
  NULL
}
E <- function()
  NULL
F <- function() {
  NULL
  NULL
}
G <- function() invisible(
)
H <- function() { invisible(
) }
```

And the corresponding lint table ( :heavy_check_mark: :arrow_right: throw a lint):

| `function_braces=` | A | B | C | D | E | F | G | H |
|---------------------------- | -- | -- | -- | -- | -- | -- | -- | -- |
| always |  |  | :heavy_check_mark:   | | :heavy_check_mark: | | :heavy_check_mark: | |
| not_inline | | | | | :heavy_check_mark: | | | |
| multi_line | | | | | :heavy_check_mark: | | :heavy_check_mark: | |
| never | | | | | | | | |
| _strict_ | :heavy_check_mark:  | | | :heavy_check_mark:  | | | | :heavy_check_mark: |

(PS we might consider adding any of the above as test cases if they're not yet represented)

Reply:
Mentioning this here to possibly align on the naming for such a style. If we can't come up with a name that's natural for the current `function_braces` (or `function_bodies`), we might need to think of a different parameter name to accommodate. Hope that makes sense.

Reply:
> use braces *if and only if they're syntactically required*.

What exactly do you mean by "syntactically required"? I'm confused by your examples above, `A`, `D` and `H` are fine without braces (same result)...

Reply:
> What exactly do you mean by "syntactically required"? 

i.e., "without braces, the code either wouldn't parse, or would parse incorrectly". In other words "function uses braces" if and only if "function has more than one expression" (I think that should be equivalent to `length(body(f)) > 1`)

> A, D and H are fine without braces (same result)...

Precisely why they would throw a lint (under "strict" mode, or maybe "minimalist" mode is more evocative) -- that style ditches `{` whenever possible. In A/D/H, ditching `{` doesn't impact the validity of the code --> lint